pkg/stream/connpool/msgconnpool/connpool.go

```diff
@@ -0,0 +1,449 @@
+	return p.host
+}
+
+// GetActiveClient get a avail client
+func (p *connpool) GetActiveClient(ctx context.Context) (*activeClient, types.PoolFailureReason) {
+	p.clientMux.Lock()
+	defer p.clientMux.Unlock()
+
+	n := len(p.idleClients)
+
+	// no available client
+	var (
+		c      *activeClient
+		reason types.PoolFailureReason
+	)
+
+	if n == 0 {
+		c, reason = p.newActiveClientLocked(ctx)
+		// if the reason if connection failure
+		// will automatic reconnect, so we can save the active client
+		// to avoid concurrent reconnect
+		if c != nil {
+			// should put this conn to pool
+			p.idleClients = append(p.idleClients, c)
+		}
+
+		return c, reason
+	} else {
+
+		var lastIdx = n - 1
+		var reason types.PoolFailureReason
+		c = p.idleClients[lastIdx]
+		if c == nil {
+			c, reason = p.newActiveClientLocked(ctx)
+			if reason == "" && c != nil {
+				p.idleClients[lastIdx] = c
+			}
+		}
+
+		return c, reason
+	}
+}
+
+// Destroy the pool
+func (p *connpool) Destroy() {
+	atomic.StoreUint64(&p.destroyed, 1)
+
+	p.clientMux.Lock()
+	defer p.clientMux.Unlock()
+
+	for _, c := range p.idleClients {
+		c.host.Connection.Close(api.NoFlush, api.LocalClose)
+	}
+}
+
+// types.StreamEventListener
+// types.ConnectionEventListener
+// types.StreamConnectionEventListener
+type activeClient struct {
+	pool *connpool
+	host *types.CreateConnectionData
+
+	keepAlive *keepAliveListener
+
+	reconnectState uint64 // for reconnect, connecting or notConnecting
+}
+
+func (p *connpool) newActiveClientLocked(ctx context.Context) (*activeClient, types.PoolFailureReason) {
+	// if pool is already destroyed, return
+	if atomic.LoadUint64(&p.destroyed) == 1 {
+		return nil, types.ConnectionFailure
+	}
+
+	createConnData := p.Host().CreateConnection(ctx)
+	ac := &activeClient{
+		pool: p,
+		host: &createConnData,
+	}
+
+	ac.host.Connection.AddConnectionEventListener(ac)
+
+	// first connect to dest addr, then create stream client
+	if err := ac.host.Connection.Connect(); err != nil {
+		if p.autoReconnectWhenClose && atomic.LoadUint64(&p.destroyed) == 0 {
+			// auto reconnect when the first connect failed
+			log.DefaultLogger.Warnf("[connpool] reconnect due to first connect failed %v", ac.host.Host.AddressString())
+			ac.Reconnect()
+		}
+
+		return ac, types.ConnectionFailure
+	} else {
+		if atomic.LoadUint64(&p.destroyed) == 1 {
+			// if destroyed, close the conn
+			ac.host.Connection.Close(api.NoFlush, api.LocalClose)
+			return ac, types.ConnectionFailure
+		}
+
+		for _, rf := range ac.pool.readFilters {
+			ac.host.Connection.FilterManager().AddReadFilter(rf)
+		}
+	}
+
+	// if user use connection pool without codec
+	// they can use the connection returned from activeClient.Conn()
+	if p.heartBeatCreator != nil {
+		ac.SetHeartBeater(p.heartBeatCreator())
+	}
+
+	return ac, ""
+}
+
+// Reconnect triggers connection to reconnect
+func (ac *activeClient) Reconnect() {
+	if !atomic.CompareAndSwapUint64(&ac.reconnectState, notConnecting, connecting) {
```

Comment:
Using optimistic lock may cause connection loss in some concurrent scenarios

Reply:
added another OnEvent logic, when RemoteClose happens during reconnecting, this conn will be marked as CloseWhenReconnect = true, and after reconnecting directly go to retry